### PR TITLE
feat(tree-widget): import from tui-rs-tree-widget

### PR DESCRIPTION
**Pull request by joshka**
Fixes https://github.com/tui-rs-revival/ratatui/issues/116
    (Integrate with [tui-rs-tree-widget](https://github.com/EdJoPaTo/tui-rs-tree-widget))

Simplified to just bring in the source instead of the history of tui-rs-tree.

Previous note:
~~Note. This PR grafts the entire commit history of the original repo. So when merging, please use the command line and perform a --ff-only merge rather than the github UI which rewrites every commit. This ensures that we can take future updates and see the rationale behind any historical decisions etc.~~

~~Only the last two commits in the history are relevant to actually review. The first commit (ca6004ccb53a19bf6e1e6a7acda2e7d56af59a2a) preps the repo for merging (i.e. moves the files around and deletes unnecessary files), the second commit (8f5a0ad983a5c617360ea68b1fdf9ef6e474b6b5) fixes up the modules etc. and properly merges the widget into ratatui.~~

~~I've left the MIT license in the README.md within the tree widget (but if @EdJoPaTo feels generous we can fold his copyright into the main license file instead - unsure what the right way to do this is).~~

~~Note also, I realized that I accidentally did the merge on top of a commit (762caa330d8c2a3ddecc0f255d331c83be282eb4) that just contains unit tests for the Paragraph, and I didn't see an easy way to untangle that single commit from the merge without re-doing the changes. Apologies (but I hope adding unit tests is not at all controversial and can be merged quickly in https://github.com/tui-rs-revival/ratatui/pull/156)~~

### CODE REVIEW

**Comment by EdJoPaTo** [comment]
I like seeing this approach! Not sure what the others think as it will add a lot of commits.

some commits contain PR numbers which then link to the wrong repo. Maybe that should be handled in some way too.

regarding the license: the current ratatui license states the original TUI author. Maybe it should be adapted to be ratatui authors or something like that? I am not sure about the legal things there relevant to such a change. But I would like to simplify the license so that the tree view widget has the same license as the rest of ratatui as that would be really annoying to handle multiple licenses for the project.

I think we should first handle the changes and added tests in my repo so this PR only contains the addition of the project? (See https://github.com/EdJoPaTo/tui-rs-tree-widget/issues/18#issuecomment-1520036311)

Sadly I don’t have that much time to work on code currently but I will happily review PRs!

Thanks for your efforts @joshka !

**Comment by joshka** [comment]
> I like seeing this approach! Not sure what the others think as it will add a lot of commits.

The graph of this is pretty neat:
```
❯ git log --oneline --graph
*   8f5a0ad (HEAD -> feat-tree-widget, origin/feat-tree-widget) feat(tree-widget) merge tui-rs-tree-widget history
|\
| * ca6004c (tree-widget/main) chore: ready tree_widget for import into ratatui
| * 61bc6da fix: panic on key up/down (#16)
| * 94197fa docs: fix spelling
...
| * 52e309b chore(workflow): lib has no lock file
| * 8ef8355 chore(workflow): add github action
| * b539806 Initial commit
* 762caa3 (origin/test-paragraph, test-paragraph) chore: add unit tests for Paragraph
* 37bb82e (upstream/main, origin/main, origin/HEAD, main) docs(readme): add acknowledgement section (#154)
* 047e0b7 style(readme): update project introduction in README.md (#153)
```
> 
> some commits contain PR numbers which then link to the wrong repo. Maybe that should be handled in some way too.

I'd suggest that when this goes ahead you lock the repo so the PRs are there if needed (perhaps move it to the tui-rs-revival org if you don't want it cluttering up your personal github). I'm not sure it's necessary to move the PRs all over.

> regarding the license: the current ratatui license states the original TUI author. Maybe it should be adapted to be ratatui authors or something like that? I am not sure about the legal things there relevant to such a change. But I would like to simplify the license so that the tree view widget has the same license as the rest of ratatui as that would be really annoying to handle multiple licenses for the project.

Yeah, that question is probably one for those that have interacted with the original author. Both licenses are MIT, which means the only necessary part is attribution / ensuring that the copyright remains, so not a huge deal. I figured you'd probably be OK with combining those. And yep, we probably should wrap it under the org as the copyright owner.

> I think we should first handle the changes and added tests in my repo so this PR only contains the addition of the project? (See [EdJoPaTo/tui-rs-tree-widget#18 (comment)](https://github.com/EdJoPaTo/tui-rs-tree-widget/issues/18#issuecomment-1520036311))

Happy to wait it out. No rush on my perspective. From the looks of that one PR that is outstanding that seems like something that could come before or after the move as it's fairly small and self contained - is it a breaking change for consumers of the tree widget lib, or one that just implements the same behavior and makes things configurable?

> Sadly I don’t have that much time to work on code currently but I will happily review PRs!
> 
> Thanks for your efforts @joshka !

No probs :)

**Comment by mindoodoo** [comment]
Re licensing issues, I believe the MIT license requires the original copyright to stay up.

> The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

If we wanted to add people / update the copyright [this](https://softwareengineering.stackexchange.com/questions/386582/altering-author-names-in-mit-license) post shows a good way of doing so :

```
Copyright (c) 2014 Florian Dehau
Copyright (c) 2019 <new author(s)>
```

cc: @orhun

**Comment by orhun** [comment]
I'm not sure about the legal side of things, but we can update the copyright like this:

```
Copyright (c) 2014 Florian Dehau
Copyright (c) 2023 ratatui contributors
```

Again, I'm not sure if this is the right way but this is what I did in the past for another project.

**Comment by EdJoPaTo** [comment]
I merged the open PR and improved some minor things in the tree widget repo and released a new version (v0.12). I think all open things are now addressed and we should continue looking into merging the tree widget into ratatui.

Some things I noticed in this PR:
- The project README.md file currently contains my tree widget library as third party library twice which should be removed and added to the list of built in widgets.
- Other widgets don't have their own README and instead are just linked to the example I think the README of the tree should just be dropped. The screenshot which currently is added as media/screenshot in this PR should also be just removed I think. It isn't adding much value for ratatui and people can easily run the example instead like with all the other widgets. Useful information about the Tree should instead be listed at the doccomment of the widget code so its included with docs.rs which is way more helpful.
- The example code seems to be annoyed currently as it cant find crossterm. It needs to be added in the Cargo.toml to know about the required feature
- The license topic should be discussed in general for ratatui probably in its own issue / PR? Personally I think I should be able to change the license of my code as I am the author of the code. Older versions of the code still continue to use the old LICENSE while newer versions would use the new license. I would prefer for simplicity to integrate it into the main license of ratatui rather than keeping its own license within ratatui.

**Comment by joshka** [comment]
Thanks @EdJoPaTo I'll redo the import and fix up the noted issues sometime in the next week or so.

**Comment by joshka** [comment]
>    The project README.md file currently contains my tree widget library as third party library twice which should be removed and added to the list of built in widgets.

Fixed

>    Other widgets don't have their own README and instead are just linked to the example I think the README of the tree should just be dropped. The screenshot which currently is added as media/screenshot in this PR should also be just removed I think. It isn't adding much value for ratatui and people can easily run the example instead like with all the other widgets. Useful information about the Tree should instead be listed at the doccomment of the widget code so its included with docs.rs which is way more helpful.

Removed

>    The example code seems to be annoyed currently as it cant find crossterm. It needs to be added in the Cargo.toml to know about the required feature

Fixed

>    The license topic should be discussed in general for ratatui probably in its own issue / PR? Personally I think I should be able to change the license of my code as I am the author of the code. Older versions of the code still continue to use the old LICENSE while newer versions would use the new license. I would prefer for simplicity to integrate it into the main license of ratatui rather than keeping its own license within ratatui.

Removed

**Comment by joshka** [comment]
Pushed fix to commit message to address lint (missed the colon)

**Review by EdJoPaTo — CHANGES_REQUESTED** [review]
Just to mention it: I think the LICENSE topic is not solved yet as the maintainers are currently not mentioned in the root LICENSE file. Personally I like the suggestion of @orhun but I am not sure about the legal results of this.

Other than my mentioned things it looks good! Thanks for your efforts @joshka!

**Comment by joshka** [comment]
I did a bit of research on the License issue. The MIT license allows us to relicense this however we want. We have a few options (and some combinations of them):

1. Doing the max amount of work:
- change the copyright owner on this library to either an individual or a collective of people (the ratatui developers)
- move the existing copyright into another NOTICES file, or leave it at the bottom of the existing LICENSE file
- Add your license to the same NOTICES / LICENSES file
2. Doing the minimum amount of work:
- Leave the copyright intact - we can alway relicense later based on the same logic
- Waive your license requirement to add a notice about tui-rs-tree-widget
3. Contact the original author and ask them to waive the license requirement in the same way as tui-rs-tree-widget and then sublicense without the extra copyright notice

I lean towards doing option 2 as it's the simplest option.

MIT License:

> Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software **without restriction**, including without limitation the rights
to use, copy, modify, merge, publish, distribute, **sublicense**, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:
>
> The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.
> ...

File: src/widgets/mod.rs
Comment (joshka) [reply src/widgets/mod.rs:33]: The Tree* types are re-exported below:
```rust
pub use self::tree::{Tree, TreeIdentifier, TreeItem, TreeState};
```
So these show up in the widgets as `ratatui::widgets::{Tree, TreeIdentifier, TreeItem, TreeState}` which is consitent with the way that we expose the rest of the widgets in the library.

I wasn't sure that exposing `tree::flatten::{flatten, Flattened}` and `tree::identifier::{get_without_leaf as get_identifier_without_leaf, TreeIdentifierVec}` without them being prefixed with `Tree` was a good idea (though I missed the `TreeIdentifierVec` there - which could be exported fine). My rationale is what does it look like having the struct / method in a tab completion list. Happy to change my mind on this if you think otherwise.

File: src/widgets/tree/identifier.rs
Comment (joshka) [reply src/widgets/tree/identifier.rs:6]: Caveat - I haven't used the tree widget, so I'm looking at it for the first time now.
Perhaps you could model this more explicitly as `struct BranchIdentifier` and `type LeafIdentifier = usize` .

Then `get_without_leaf` method could easily be moved to be a method implemented on `BranchIdentifier`, and rename it to `BranchIdentifier.split_leaf()` or something like that (not sure what would be idiomatic).

My take on performance is that until you measure it, it doesn't really exist. The sorts of projects that render data in a TUI would have approaches to segment a large tree (perhaps) rather than rendering something so huge that cloning identifiers is an issue (this is an assumption - possibly I'm wrong). Either way, I'd suggest pick a node count that represents a realistic upper bound on tree nodes and benchmark that to get a feel for whether you're talking nano seconds or seconds for a clone (or some number in between), and then worrying about it when it's not fast enough for actual use cases that you encounter.

File: src/widgets/mod.rs
Comment (EdJoPaTo) [reply src/widgets/mod.rs:33]: I think when searching for tree thingies it's also quite easy to find `ratatui::widgets::tree::` via tab completion. Personally I would avoid exposing the same things on multiple places as it might end up with different code for the same thing.

Personally I would just go with exposing the tree module and don't reexport them again. Then nothing can be missed accidentally for example and its a bit more grouped rather than exposing a bunch of tree related stuff on the widget module.

Comment (joshka) [reply src/widgets/mod.rs:33]: > Personally I would just go with exposing the tree module and don't reexport them again. Then nothing can be missed accidentally for example and its a bit more grouped rather than exposing a bunch of tree related stuff on the widget module.

I agree with your logic on this.

File: src/widgets/tree/mod.rs
Comment (joshka) [reply src/widgets/tree/mod.rs:2]: Agreed

**Comment by EdJoPaTo** [comment]
What's the current state here? What's needed for a merge?

Personally I would like to see this widget in ratatui rather than having it as a feature on my crate to keep things simpler. But it seems like people want to use the tree widget and would like to have ratatui support for it so maybe its faster to just add it as a feature?

**Comment by joshka** [comment]
Regarding the license, we've added "The ratatui developers" to the copyright in the license file, which so long as you're fine with that approach, you're included in that now.

**Comment by joshka** [comment]
Before merging this, I'd like to add a some more tests, and to understand how the identifier code works a bit better.

**Comment by EdJoPaTo** [comment]
> I need to re-import it. I think it's probably not worth the effort to do the merge unrelated history approach I'd taken previously as it complicates things a bit and makes it difficult to rebase if necessary before merging.

I would like to add myself as an author to the commit there so the single commit at least knows that it's originally from me / the original repo. I assume I would be listed then as author on this repo then?

I think it simplifies a lot to just squash stuff into the single commit. 👍

**Comment by joshka** [comment]
Done

File: README.md
Comment (EdJoPaTo) [inline README.md:201]: This is unrelated to the tree widget. Was this accidentally added or just saw it missing and added it?

File: README.md
Comment (EdJoPaTo) [inline README.md:205]: The exports are from the submodule and not directly

```suggestion
* [Tree](https://docs.rs/ratatui/latest/ratatui/widgets/tree/struct.Tree.html)
```

**Review by EdJoPaTo — CHANGES_REQUESTED** [review]

File: src/widgets/tree/mod.rs
Comment (EdJoPaTo) [inline src/widgets/tree/mod.rs:6]: Why are these star imports needed? They don't seem specific and are therefor a bit shady to me. Explicit imports seem like a better idea to me and personally I would just import stuff like Text, Block and so on explicitly.

Guess the prelude in itself is highly based on preference and personally I try to follow all the pedantic clippy lints including [wildcard_imports](https://rust-lang.github.io/rust-clippy/master/index.html#/wildcard_imports) (which allows prelude::* by default but I don't really see the point in less explicit imports for a specific name like prelude)

File: src/widgets/tree/mod.rs
Comment (EdJoPaTo) [reply src/widgets/tree/mod.rs]: done in #332

**Comment by fujiapple852** [comment]
@EdJoPaTo do you think it is feasible to use inside a `ratatui::widgets::table:Cell`?

**Comment by EdJoPaTo** [comment]
> @EdJoPaTo do you think it is feasible to use this inside a `ratatui::widgets::table:Cell`?

I think this should be discussed in a separate issue as this one is specific to bringing the tree widget into ratatui. As this is not yet merged it should be in the tree widget repo.

Currently I can not really answer the question as I don't really know the requirements for something being a Cell. I would need to look into it at another time which would also be helpful on a separate issue. Maybe other people know more about it and can join into the discussion then.

**Comment by joshka** [comment]
I know this has been stale for a while - mostly because I got busy doing other things. I think what we'd like to do with this is to move it into the ratatui-org but keep it as a separate project for now. The idea being that it's something that we support, but also something that has a different release schedule etc. Closing this out for now.

**Pull request closed without merging**
